GitCommands/CommitDataManager.cs

```diff
@@ -15,17 +16,20 @@ public interface ICommitDataManager
         /// <param name="data">Data produced by a <c>git log</c> or <c>git show</c> command where <c>--format</c>
         /// was provided the string <see cref="CommitDataManager.LogFormat"/>.</param>
         /// <returns>CommitData object populated with parsed info from git string.</returns>
-        CommitData CreateFromFormatedData(string data);
+        [NotNull]
```

Comment:
I kind of expect to be laughed at for a noob question, but - what do we get with these attribute decorations?

Reply:
At a minimum, these attributes serve as documentation of intent around whether values can or can not be null.

If you use R#, then they enable warnings about potential NullReferenceExceptions, (eg. passing null to a method that says it cannot be null, or using a nullable return value without checking first), and redundant code (eg. checking something for null that can never be null).

The act of annotating methods also makes you think a little more about a method's behaviour, or a field's invariants.

Hopefully in C# 8 support for [nullable reference types](https://blogs.msdn.microsoft.com/dotnet/2017/11/15/nullable-reference-types-in-csharp/) will land, then the attributes can be converted to actual language constructs.

That is:

```c#
// before

[Nullable] private Bar _bar;
[NotNull] private Baz _baz;

[Nullable] public string Foo([NotNull] string s) { ... }

// after

private Bar? _bar;
private Baz _baz;

public string? Foo(string s) { ... }
```

Annotating now will help with this migration later.

The above covers `[Nullable]` and `[NotNull]` annotations, which are the most common. There are also:

- `[Pure]` for a method with no observable side effects (can be hoisted, and return value should not be discarded)
- `[InstantHandle]` for `IEnumerable<T>` and delegate parameters that are fully evaluated before the method returns
- `[ItemNotNull]` and `[ItemNullable]` for collection and `Task<T>` types
- `[ContractAnnotation]` that specifies how inputs and outputs relate to each other (eg. if the input is `null` the method will return `false`, allowing R# to flag something like `if (!FalseIfNull(notNullValue)) { /* unreachable code */ }` and so forth)

You can read more about them [here](https://www.jetbrains.com/help/resharper/Code_Analysis__Code_Annotations.html).

All these attributes are defined conditionally meaning they do not end up in the compiled binary, so there's no overhead to using them.